esphome/components/gsm/gsm.cpp

```diff
@@ -0,0 +1,232 @@
+#include "gsm.h"
+#include "esphome/core/log.h"
+#include <string.h>
+
+namespace esphome {
+namespace gsm {
+
+static const char* TAG = "gsm";
+
+const char ASCII_CR = 0x0D;
+const char ASCII_LF = 0x0A;
+
+void GSMComponent::setup() {}
+
+void GSMComponent::update() {
+  if (this->whatch_dog_++ == 2)
+    this->state_ = STATE_INIT;
+
+  if (state_ == STATE_INIT) {
+    if (this->registered_ && this->send_pending_) {
+      this->send_cmd_("AT+CMGS=\"" + this->recipient_ + "\"");
+      this->state_ = STATE_SENDINGSMS1;
+    } else {
+      this->send_cmd_("AT");
+      this->state_ = STATE_CHECK_AT;
+      this->expect_ack_ = true;
+    }
+  }
+  if (state_ == STATE_RECEIVEDSMS) {
+    // Serial Buffer should have flushed.
+    // Send cmd to delete received sms
+    char delete_cmd[10];
+    sprintf(delete_cmd, "AT+CMGD=%d", this->parse_index_);
+    this->send_cmd_(delete_cmd);
+    this->state_ = STATE_CHECK_SMS;
+    this->expect_ack_ = true;
+  }
+}
+
+void GSMComponent::send_cmd_(std::string message) {
+  ESP_LOGV(TAG, "S: %s - %d", message.c_str(), this->state_);
+  this->whatch_dog_ = 0;
+  this->write_str(message.c_str());
+  this->write_byte(ASCII_LF);
+}
+
+void GSMComponent::parse_cmd_(char* message) {
+  ESP_LOGV(TAG, "R: %s - %d", message, this->state_);
+
+  if (strlen(message) == 0)
+    return;  // Empty line;
+
+  if (this->expect_ack_) {
+    bool ok = strcmp(message, "OK") == 0;
+    this->expect_ack_ = false;
+    if (!ok) {
+      if (this->state_ == STATE_CHECK_AT && strcmp(message, "AT") == 0) {
+        // Expected ack but AT echo received
+        this->state_ = STATE_DISABLE_ECHO;
+        this->expect_ack_ = true;
+      } else {
+        ESP_LOGE(TAG, "Not ack. %d %s", this->state_, message);
+        this->state_ = STATE_IDLE;  // Let it timeout
+        return;
+      }
+    }
+  }
+
+  switch (this->state_) {
+    case STATE_DISABLE_ECHO:
+      send_cmd_("ATE0");
+      this->state_ = STATE_CHECK_AT;
+      this->expect_ack_ = true;
+      break;
+    case STATE_CHECK_AT:
+      send_cmd_("AT+CMGF=1");
+      this->state_ = STATE_CREG;
+      this->expect_ack_ = true;
+      break;
+    case STATE_CREG:
+      send_cmd_("AT+CREG?");
+      this->state_ = STATE_CREGWAIT;
+      break;
+    case STATE_CREGWAIT: {
+      // Response: "+CREG: 0,1" -- the one there means registered ok
+      //           "+CREG: -,-" means not registered ok
+      message[6] = 0;
+      bool registered = strcmp(message, "+CREG:") == 0 && message[9] == '1';
+      if (registered) {
+        if (!this->registered_)
+          ESP_LOGI(TAG, "Registered OK");
```

Comment:
```suggestion
          ESP_LOGD(TAG, "Registered OK");
```

Sounds more like a debug message to me.